### DISCUSSION TITLE
Does Shairport Built with Airplay 2 Support the Ability to Fallback to Airplay 1

### DISCUSSION THREAD

**Discussion by PablaV** [mikebrady/shairport-sync#1691] [General]
I am using Shairport built with Airplay 2 and it works fine when using Airplay 2 supported players. I am trying to use a addon from Home Assistant called Music Assistant 2.0, it has the ability to stream via Airplay 1 just not Airplay 2. To my understanding and testing, Airplay 2 has reverse compatibility with Airplay 1, meaning if a Airplay 1 device streams to the Airplay 2 receiver it can still play the audio. In my testing Music Assistant is able to play to me Airplay 2 devices no problem. However it cannot stream to Shairport, before I dig deeper into this I want to know if Shairport supports the backwards compatibility in the first place. I talked beifly with the Music Assistant devs and they believe this may be a Shairport issue since Music Assistant is able to stream to other Airplay 2 devices without issues. 

PS I know you can build Shairport using Airplay 1 but I prefer Airplay 2

**Comment by PablaV** [comment]
Heres a snipped of some logs when I try to play something from Music Assistant to Shairport
```
Jun 14 11:27:21 raspberrypi shairport-sync:          0.603771302 "rtsp.c:747" mutex_lock "&conns_lock".
Jun 14 11:27:21 raspberrypi shairport-sync:          0.000154010 "common.c:1565" mutex_unlock "&conns_lock" at "rtsp.c:758".
Jun 14 11:27:21 raspberrypi shairport-sync:          0.000169844 "rtsp.c:5564" Connection 9: New connection from 192.168.8.85:35433 to self at 192.168.8.27:7000.
Jun 14 11:27:21 raspberrypi shairport-sync:          0.000208073 "rtsp.c:5580" Successfully created RTSP receiver thread 9.
Jun 14 11:27:21 raspberrypi shairport-sync:          0.000058542 "rtsp.c:687" mutex_lock "&conns_lock".
Jun 14 11:27:21 raspberrypi shairport-sync:          0.000054427 "common.c:1565" mutex_unlock "&conns_lock" at "rtsp.c:709".
Jun 14 11:27:21 raspberrypi shairport-sync:          0.100837916 "rtsp.c:808" msg_init message 5
Jun 14 11:27:21 raspberrypi shairport-sync:          0.000147969 "rtsp.c:917" RTSP Message Received: "POST /auth-setup RTSP/1.0".
Jun 14 11:27:21 raspberrypi shairport-sync:          0.000067136 "rtsp.c:948"     Content-Type: application/octet-stream.
Jun 14 11:27:21 raspberrypi shairport-sync:          0.000054166 "rtsp.c:948"     Content-Length: 33.
Jun 14 11:27:21 raspberrypi shairport-sync:          0.000052136 "rtsp.c:948"     CSeq: 1.
Jun 14 11:27:21 raspberrypi shairport-sync:          0.000051771 "rtsp.c:948"     User-Agent: iTunes/7.6.2 (Windows; N;).
Jun 14 11:27:21 raspberrypi shairport-sync:          0.000052395 "rtsp.c:948"     Client-Instance: 2e9a538fc86476e2.
Jun 14 11:27:21 raspberrypi shairport-sync:          0.000053177 "rtsp.c:948"     DACP-ID: 1A2B3D4EA1B2C3D4.
Jun 14 11:27:21 raspberrypi shairport-sync:          0.000053178 "rtsp.c:948"     Active-Remote: 2424290219.
Jun 14 11:27:21 raspberrypi shairport-sync:          0.000062500 "rtsp.c:808" msg_init message 6
Jun 14 11:27:21 raspberrypi shairport-sync:          0.000049947 "rtsp.c:5149" Connection 9: Received an RTSP Packet of type "POST":
Jun 14 11:27:21 raspberrypi shairport-sync:          0.000050521 "rtsp.c:5151"   Type: "Content-Type", content: "application/octet-stream"
Jun 14 11:27:21 raspberrypi shairport-sync:          0.000049115 "rtsp.c:5151"   Type: "Content-Length", content: "33"
Jun 14 11:27:21 raspberrypi shairport-sync:          0.000046354 "rtsp.c:5151"   Type: "CSeq", content: "1"
Jun 14 11:27:21 raspberrypi shairport-sync:          0.000046042 "rtsp.c:5151"   Type: "User-Agent", content: "iTunes/7.6.2 (Windows; N;)"
Jun 14 11:27:21 raspberrypi shairport-sync:          0.000048229 "rtsp.c:5151"   Type: "Client-Instance", content: "2e9a538fc86476e2"
Jun 14 11:27:21 raspberrypi shairport-sync:          0.000047344 "rtsp.c:5151"   Type: "DACP-ID", content: "1A2B3D4EA1B2C3D4"
Jun 14 11:27:21 raspberrypi shairport-sync:          0.000046093 "rtsp.c:5151"   Type: "Active-Remote", content: "2424290219"
Jun 14 11:27:21 raspberrypi shairport-sync:          0.000047344 "rtsp.c:5151"   No Content Plist. Content length: 33.
Jun 14 11:27:21 raspberrypi shairport-sync:          0.000058906 "rtsp.c:2568" Connection 9: Unhandled POST /auth-setup Content-Length 33
Jun 14 11:27:21 raspberrypi shairport-sync:          0.000048907 "rtsp.c:2570" POST request
Jun 14 11:27:21 raspberrypi shairport-sync:          0.000046093 "rtsp.c:2570"   Type: "Content-Type", content: "application/octet-stream"
Jun 14 11:27:21 raspberrypi shairport-sync:          0.000048073 "rtsp.c:2570"   Type: "Content-Length", content: "33"
Jun 14 11:27:21 raspberrypi shairport-sync:          0.000048021 "rtsp.c:2570"   Type: "CSeq", content: "1"
Jun 14 11:27:21 raspberrypi shairport-sync:          0.000047761 "rtsp.c:2570"   Type: "User-Agent", content: "iTunes/7.6.2 (Windows; N;)"
Jun 14 11:27:21 raspberrypi shairport-sync:          0.000047395 "rtsp.c:2570"   Type: "Client-Instance", content: "2e9a538fc86476e2"
Jun 14 11:27:21 raspberrypi shairport-sync:          0.000046250 "rtsp.c:2570"   Type: "DACP-ID", content: "1A2B3D4EA1B2C3D4"
Jun 14 11:27:21 raspberrypi shairport-sync:          0.000046146 "rtsp.c:2570"   Type: "Active-Remote", content: "2424290219"
Jun 14 11:27:21 raspberrypi shairport-sync:          0.000121511 "rtsp.c:2570"   No Content Plist. Content length: 33.
Jun 14 11:27:21 raspberrypi shairport-sync:          0.000048593 "rtsp.c:5210" Connection 9: RTSP Response:
Jun 14 11:27:21 raspberrypi shairport-sync:          0.000049011 "rtsp.c:5211"   Response Code: 200.
Jun 14 11:27:21 raspberrypi shairport-sync:          0.000046510 "rtsp.c:5211"   Type: "CSeq", content: "1"
Jun 14 11:27:21 raspberrypi shairport-sync:          0.000046198 "rtsp.c:5211"   Type: "Server", content: "AirTunes/366.0"
Jun 14 11:27:21 raspberrypi shairport-sync:          0.000046511 "rtsp.c:5211"   No Content Plist. Content length: 0.
Jun 14 11:27:21 raspberrypi shairport-sync:          0.000246041 "rtsp.c:893" msg_free freed message 6
Jun 14 11:27:21 raspberrypi shairport-sync:          0.000074636 "rtsp.c:893" msg_free freed message 5
Jun 14 11:27:22 raspberrypi shairport-sync:          0.420722968 "rtsp.c:1353" Connection 9: Connection closed by client.
Jun 14 11:27:22 raspberrypi shairport-sync:          0.000137188 "rtsp.c:5282" Connection 9: Terminate RTSP connection.
Jun 14 11:27:22 raspberrypi shairport-sync:          0.000057031 "rtsp.c:4985" Connection 9: rtsp_conversation_thread_func_cleanup_function called.
Jun 14 11:27:22 raspberrypi shairport-sync:          0.000050729 "rtsp.c:2657" Connection 9: TEARDOWN unspecified stream connection.
Jun 14 11:27:22 raspberrypi shairport-sync:          0.000050104 "rtsp.c:4999" Connection 9: terminating  -- closing timing, control and audio sockets...
Jun 14 11:27:22 raspberrypi shairport-sync:          0.000048334 "rtsp.c:5020" Connection 9: terminating -- closing RTSP connection socket 12: from 192.168.8.85:35433 to self at 192.168.8.27:7000.
Jun 14 11:27:22 raspberrypi shairport-sync:          0.000204218 "rtsp.c:5073" Cancel watchdog thread.
Jun 14 11:27:22 raspberrypi shairport-sync:          0.000081407 "rtsp.c:5075" Join watchdog thread.
Jun 14 11:27:22 raspberrypi shairport-sync:          0.000201198 "rtsp.c:635" Connection 9: Watchdog Exit.
Jun 14 11:27:22 raspberrypi shairport-sync:          0.000415937 "rtsp.c:5077" Delete watchdog mutex.
Jun 14 11:27:22 raspberrypi shairport-sync:          0.000060833 "rtsp.c:5083" Connection 9: Closed.
Jun 14 11:27:22 raspberrypi shairport-sync:          0.870790260 "dacp.c:505" mutex_lock "&dacp_server_information_lock".
Jun 14 11:27:24 raspberrypi shairport-sync:          2.000198281 "dacp.c:505" mutex_lock "&dacp_server_information_lock".
Jun 14 11:27:26 raspberrypi shairport-sync:          2.000270676 "dacp.c:505" mutex_lock "&dacp_server_information_lock".
Jun 14 11:27:28 raspberrypi shairport-sync:          2.000241302 "dacp.c:505" mutex_lock "&dacp_server_information_lock".
Jun 14 11:27:30 raspberrypi shairport-sync:          2.000198020 "dacp.c:505" mutex_lock "&dacp_server_information_lock".
Jun 14 11:27:32 raspberrypi shairport-sync:          2.000239114 "dacp.c:505" mutex_lock "&dacp_server_information_lock".
Jun 14 11:27:34 raspberrypi shairport-sync:          2.000241874 "dacp.c:505" mutex_lock "&dacp_server_information_lock".
Jun 14 11:27:36 raspberrypi shairport-sync:          2.000224426 "dacp.c:505" mutex_lock "&dacp_server_information_lock".
```

**Comment by mikebrady** [comment]
Thanks for the interesting posts. It looks like iTunes for Windows is trying to establish a connection to Shairport Sync -- is that right?

**Reply by PablaV** [reply]
It's not a windows device per say, it's a add on running in Home Assistant called Music Assistant. I'm assuming it plays to airplay devices using a similar fashion as windows iTunes does.

**Reply by mikebrady** [reply]
Thanks. The issue is that Windows iTunes uses some kind of authentication that has not been figured out for Shairport Sync. If Music Assistant has figured it out, that would be very interesting indeed. Is it possible that Music Assistant is actually getting iTunes to do the authentication? (Pardon my ignorance, please!)

**Reply by PablaV** [reply]
I got in touch with the Music Assistant dev to see if he can chime in here!

**Reply by PablaV** [reply]
From a dev named Marcel:  

"Hi, well, Music Assistant is using Squeeze2Raop under the hood, which is based on slimproto nd is created by philippe44. I'm pretty sure Mike will know that project as its the same as used by LMS. So basically if it works with LMS, it will also work with MA.
<https://github.com/philippe44/LMS-Raop>
"

**Reply by mikebrady** [reply]
Super, thanks. I’ll see if I can make some sense of it. Shairport Sync does have AirPlay 1 backward compatibility, so let’s hope for the best. It make take some time to figure it all out though.

**Reply by philippe44** [reply]
What authentication are you referring to? I would be surprised that I have figured out something you did not already do 😃

**Reply by mikebrady** [reply]
Hi @philippe44. Forgive me -- I may have this completely misunderstood, but it seems that Music Assistant initiates an authorisation setup:
```
0.000147969 "rtsp.c:917" RTSP Message Received: "POST /auth-setup RTSP/1.0".
```
Does that seem right, and if so, what should SPS do? Apologies if I'm not making a lot of sense...

**Reply by philippe44** [reply]
Yes, AFAIR a device that has an "et" field of "4" in mDNS is a mFI and requests a pubkey to be sent in an auth-setup request. I think it's all it requires so I build a secret, get a pubkey out of it and send it. I don't remember using it after then but I might be wrong and would need to reread that code ... that was a while ago but indeed I remember something was funky about that

**Reply by mikebrady** [reply]
Super -- I'll play with that, thanks. 🙏 (I might be coming back to you!)

**Reply by philippe44** [reply]
In fact, in my LMS-raop repo (this is squeeze2raop) I put a note in CHANGELOG about that in 0.4.4.0 and 0.4.4.1. The note is not very clear, sorry. Sometimes I just write these really quick and it does not always make sense when read later

**Comment by PablaV** [comment]
Just a little more info, I built another instance of shairport this time using airplay 1 and I was successfully able to play to it from Music Assistant

**Reply by mikebrady** [reply]
Hi there. An update has been pushed to the `development` branch, which might address this issue. If you got a chance to test it and report back, it would be appreciated.

**Reply by PablaV** [reply]
Gave the dev branch a try and Music Assistant can finally connect to Shairport! A few small things I noticed was even though I have the name set to show up as "Shairport" on airplay it shows up as "raspberry pi" not that big of a deal. Second thing I noticed is that Owntone now throws errors for artwork even though the artwork is coming through fine, again not a big deal I can ignore the errors.

```
Jul  6 11:03:27 raspberrypi owntone[31468]: [2023-07-06 11:03:27] [  LOG]  artwork: Source 'stream' returned an error for 'shair'
Jul  6 11:03:27 raspberrypi owntone[31468]: [2023-07-06 11:03:27] [  LOG]  artwork: Source 'stream' returned an error for 'shair'
Jul  6 11:03:27 raspberrypi owntone[31468]: [2023-07-06 11:03:27] [  LOG]  artwork: Source 'stream' returned an error for 'shair'
Jul  6 11:03:27 raspberrypi owntone[31468]: [2023-07-06 11:03:27] [  LOG]  artwork: Source 'stream' returned an error for 'shair'
Jul  6 11:03:30 raspberrypi owntone[31468]: [2023-07-06 11:03:30] [  LOG]  artwork: Source 'stream' returned an error for 'shair'
Jul  6 11:03:30 raspberrypi owntone[31468]: [2023-07-06 11:03:30] [  LOG]  artwork: Source 'stream' returned an error for 'shair'
Jul  6 11:03:35 raspberrypi owntone[31468]: [2023-07-06 11:03:35] [  LOG]  artwork: Source 'stream' returned an error for 'shair'
Jul  6 11:03:35 raspberrypi owntone[31468]: [2023-07-06 11:03:35] [  LOG]  artwork: Source 'stream' returned an error for 'shair'
```

**Reply by mikebrady** [reply]
Super -- I'm glad it's working.

Regarding the other phenomena: 
1. Shairport Sync will use the `hostname` of the computer as the service name by default. You can change that in the configuration file or on the command line.
2. I don't know about the OwnTone stuff, but those guys are very responsive.

The service name and artwork settings are no different (well, they *should* not be different) to the Classic (AirPlay 1) version...

Big thanks to @philippe44 for putting me own the right track here!

**Reply by PablaV** [reply]
Thanks Mike and @philippe44 for helping out :)

**Comment by bverkron** [comment]
> Just a little more info, I built another instance of shairport this time using airplay 1 and I was successfully able to play to it from Music Assistant

@PablaV How do you set the Shairport instance to use Airplay 1 only? I would like to force mine to AP 1 so that I get bit perfect playback instead of resampling, at least that's what I'd like to attempt.

Edit: Never mind. I see in the docker image read me it says to use the `:classic` tag. I'm using docker so I'll try that.
